File: ios/Packages/AppFeatures/Sources/HomeFeature/HomeViewModel.swift
import ConfigKit
import CoreKit
import Foundation
import Observation

/// Renders the server-composed Home layout through the native section catalog
/// (ADR-0011). Unknown section types are dropped here — never an error.
@MainActor
@Observable
public final class HomeViewModel {
    /// Section catalog v1 — the native components this app version can render.
    public static let supportedSections: Set<String> = [
        "ambient_header", "prayer_hero", "ask_ai", "quick_actions",
    ]

    public private(set) var sections: [HomeLayout.Section] = []
    public private(set) var askEnabled = false

    private let config: ConfigService
    private let appVersion: String

    public init(config: ConfigService, appVersion: String) {
        self.config = config
        self.appVersion = appVersion
    }

    /// Bundled default layout — mirrors backend seed; used until a server
    /// layout is cached (ADR-0011: bundled defaults are offline fallbacks).
    static let fallbackLayout = HomeLayout(version: 0, sections: [
        .init(id: "header", type: "ambient_header", props: [:]),
        .init(id: "prayer", type: "prayer_hero", props: ["show_timeline": .bool(true)]),
        .init(id: "ask", type: "ask_ai", props: ["state": .string("coming_soon")]),
        .init(id: "quick", type: "quick_actions", props: [:]),
    ])

    public func load() async {
        let layout = await config.current.homeLayout ?? Self.fallbackLayout
        sections = layout.renderableSections(supported: Self.supportedSections)
        askEnabled = await config.isEnabled("module.ai_ask", appVersion: appVersion)
    }

    public func refresh(locales: [String]) async {
        await config.refresh(locales: locales)
        await load()
    }
}

File: android/feature/home/src/main/kotlin/com/fatwabot/feature/home/HomeViewModel.kt
package com.fatwabot.feature.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.fatwabot.core.common.HomeLayout
import com.fatwabot.core.config.ConfigService
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Resolves the server Home layout + flags (ADR-0011) — mirror of iOS HomeViewModel.
 * Unknown section types are dropped by the renderer, never here.
 */
@HiltViewModel
class HomeViewModel @Inject constructor(
    private val config: ConfigService,
) : ViewModel() {

    data class UiState(
        val layout: HomeLayout? = null,
        val askEnabled: Boolean = false,
    )

    private val _state = MutableStateFlow(UiState(layout = config.homeLayout()))
    val state: StateFlow<UiState> = _state.asStateFlow()

    fun load(appVersion: String) {
        _state.value = UiState(
            layout = config.homeLayout(),
            askEnabled = config.isEnabled("module.ai_ask", appVersion),
        )
    }

    fun refresh(appVersion: String, locales: List<String>) {
        viewModelScope.launch {
            config.refresh(locales)
            load(appVersion)
        }
    }
}
